### DISCUSSION TITLE
book deltas vs snapshots

### DISCUSSION THREAD

**Discussion by doublier1** [nautechsystems/nautilus_trader#1211] [General]
I have a question on how to handle book deltas vs snapshots. Traditionally the way this works is that the exchanges recommend you subscribe to deltas, then get an initial snapshot. Only then do you start applying deltas to the book (making sure to drop any with a sequence number less than that on the initial snapshot). What I find in Nautilus (on Binance) is a bit odd. If I subscribe to a snapshot and deltas, then the books that are generated by each callback are substantially different from each other. If I just use the deltas and build a book, I'm never actually guaranteed to know what the actual best bid/offer is, as the starting point is random. If I subscribe to only snapshots, then the limitation is that you only get aggregated events each interval_ms, so you potentially are reacting slower than other participants when then is a material update to the orderbook. What is the recommended way to handle this?

**Comment by cjdsellers** [comment]
Cross posting from Discord

> Hi @lawman 
>
> Thanks for your question, and you're quite correct in your understanding of how snapshots + deltas should work. I believe this is how Nautilus is implemented for Binance by default, if you refer to the source here: https://github.com/nautechsystems/nautilus_trader/blob/develop/nautilus_trader/adapters/binance/common/data.py#L292
> 
> Deltas are first subscribed to and start buffering, when the snapshot is received we then apply all deltas from the next update ID which the snapshot doesn't have, discarding the others (basically as you suggest).
> 
> It's important to remember that initializing books like this is specific to each adapter, as implementations may vary depending on the API the venue offers. I did a quick test on Binance by pretty printing the book (book.pprint(20) where the 20 is the number of levels (3 is default), and it looked exactly in line with the order book seen in the GUI (Nautilus won't just start applying deltas to an empty book, by default).
> 
> Could you share in a bit more detail how you're using the different order book methods, maybe even a code snippet showing what you're subscribing for, are you initiailizing and building your own book? (you don't actually need to, the DataEngine will handle this) and you can either wait for the OrderBook snapshots to arrive in your strategy handler, or pull the book out of the cache whenever you like. Of course you may be interested in the individual deltas themselves too for any reason

**Comment by cjdsellers** [comment]
This turned out to be a bug when requesting for full depth books, an initial snapshot was not being requested. This has been changed to request an initial 1000 level snapshot, before applying deltas (which works for both the Futures and Spot APIs) 970d5511a80a1e883f0bc225b5a253727f5b2337.
